crates/scheduler/src/allocator.rs

```diff
@@ -181,69 +213,123 @@ impl<S> Stream for GreedyOfferAggregator<S>
-        if let Some(max) = this.max_offers
-            && *this.offers_received >= *max
-        {
-            tracing::debug!("Reached maximum offer limit");
-            return Poll::Ready(this.best_offer.take());
+        if let Some(max) = this.max_offers {
+            if *this.offers_received >= *max {
+                tracing::debug!("Reached maximum offer limit");
+                let result = std::mem::take(this.selected);
+                *this.returned = true;
+                return Poll::Ready(Some(result));
+            }
         }
 
         // Poll the deadline timer
         if this.deadline.as_mut().poll(cx).is_ready() {
             tracing::debug!("Deadline reached");
-            return Poll::Ready(this.best_offer.take());
+            let result = std::mem::take(this.selected);
+            *this.returned = true;
+            return Poll::Ready(Some(result));
         }
 
         // Poll for new offers
         loop {
             match this.stream.as_mut().poll_next(cx) {
                 Poll::Ready(Some((peer_id, offer))) => {
                     *this.offers_received += 1;
+                    let mut changed = false;
 
-                    // Update best offer if this one is better
-                    let should_update_timer = match this.best_offer {
-                        Some((_, current_best)) if current_best.price <= offer.price => {
-                            // Keep current best
-                            false
+                    if *this.diversity {
+                        if let Some(&idx) = this.index_by_peer.get(&peer_id) {
+                            // Already have an entry for this peer; keep the cheaper one.
+                            if offer.price < this.selected[idx].1.price {
+                                this.selected[idx] = (peer_id, offer.clone());
+                                changed = true;
+                            }
+                        } else {
+                            // New peer
+                            if this.selected.len() < *this.desired {
+                                this.selected.push((peer_id, offer.clone()));
+                                changed = true;
+                            } else if let Some((max_idx, max_price)) =
+                                current_max_price_index(this.selected)
+                            {
+                                if offer.price < max_price {
+                                    let removed = std::mem::replace(
+                                        &mut this.selected[max_idx],
+                                        (peer_id, offer.clone()),
+                                    );
+                                    // Update index map for removed peer
+                                    this.index_by_peer.remove(&removed.0);
+                                    changed = true;
+                                }
+                            }
```

Comment:
NIT: Do you want to refactor this into a seperate function, because you will do the same below with only one line changed.